From 854d7ebba4fb2e1ed0bf724ee6926824ca3054a6 Mon Sep 17 00:00:00 2001
From: Meng Li <Meng.Li@windriver.com>
Date: Fri, 17 Dec 2021 12:17:39 +0800
Subject: [PATCH 2/2] firmware: stratix10-svc: release mutex sdm_lock after
 processing command

commit 94415527742e10e8935eda52e29d11095407cb6a upstream

When loading intel-fcs.ko kernel module, system hangs up. This issue
is introduced by sdk commit fc23818c8dfa("HSD #14015013554: firmware:
stratix10-svc: support up to 4 SVC clients "). This commit intends to declear
that stratix10-svc driver can support 4 SVC clients by allocating their own
channel to communicate with ATF, and use a  mutex to control thread
synchronization. But the mutex sdm_lock is not released after processing
command, and cause a deadlock issue. So, add the mutex release code after
processing command.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
[DP: backport of 94415527742e from v5.10/standard/intel-sdk-5.10/intel-socfpga branch of linux-yocto]
Signed-off-by: Dragos-Marian Panait <dragos.panait@windriver.com>
---
 drivers/firmware/stratix10-svc.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/firmware/stratix10-svc.c b/drivers/firmware/stratix10-svc.c
index 85a8ec247a43..861cb2464cf6 100644
--- a/drivers/firmware/stratix10-svc.c
+++ b/drivers/firmware/stratix10-svc.c
@@ -1034,6 +1034,8 @@ static int svc_normal_to_secure_thread(void *data)
 			break;
 
 		}
+		mutex_unlock(ctrl->sdm_lock);
+		sdm_lock_owned = false;
 	}
 	pr_debug("%s: %s: Exit thread\n", __func__, chan->name);
 	if (sdm_lock_owned == true)
-- 
2.34.1

